src/snowflake/snowpark/column.py

```diff
@@ -734,21 +892,65 @@ def within_group(
             ----------------
             <BLANKLINE>
         """
-        return Column(
-            WithinGroup(
-                self._expression,
-                [
-                    _to_col_if_str(col, "within_group")._expression
-                    for col in parse_positional_args_to_list(*cols)
-                ],
-            )
+        prop = lambda ast: ast.sp_column_within_group
+        ast = Column._create_ast(
+            property = prop,
+            copy_messages = {"col": self._ast},
+            assign_fields = {"variadic": (len(cols) > 1 or not isinstance(cols[0], (list, tuple, set)))}
         )
 
+        order_by_cols = []
+        for col in parse_positional_args_to_list(*cols):
+            if isinstance(col, Column):
+                order_by_cols.append(col)
+                prop(ast).cols.append(col._ast)
+            elif isinstance(col, str):
+                col_ast = prop(ast).cols.add()
+                new_col = Column(col, ast = col_ast)
+                col_ast.sp_column.name = new_col.get_name()
+                order_by_cols.append(new_col)
+            else:
+                raise TypeError(
+                    f"'WITHIN_GROUP' expected Column or str, got: {type(col)}"
+                )
+
+        return Column(WithinGroup(self._expression, order_by_cols), ast = ast)
+
     def _named(self) -> NamedExpression:
         if isinstance(self._expression, NamedExpression):
             return self._expression
         else:
             return UnresolvedAlias(self._expression)
+    
+    @staticmethod
+    def _create_ast(property: Optional[Callable] = None,
+                    assign_fields: Dict[str, Any] = {},
+                    copy_messages: Dict[str, Any] = {},
+                    fill_col_asts: Dict[str, ColumnOrLiteral] = {}
+        ) -> proto.SpColumnExpr:
+        ast = proto.SpColumnExpr()
+        if property is not None:
+            for attr, value in assign_fields.items():
+                setattr(property(ast), attr, value)
+            for attr, messg in copy_messages.items():
+                getattr(property(ast), attr).CopyFrom(messg)
+            for attr, other in fill_col_asts.items():
+                Column._fill_ast(getattr(property(ast), attr), other)
+        return ast
+    
+    @classmethod
+    def _fill_ast(cls, ast: proto.SpColumnExpr, value: ColumnOrLiteral) -> None:
+        """
+        Copy from a Column object's AST, or copy a literal value into an AST expression.
```

Comment:
please explain params, use Google Style (you can configure in PyCharm) to explain params.